audiotrack.go

```diff
@@ -0,0 +1,308 @@
+	// and writes them to pcmWriter. If no resampling is needed, we directly point resampledPCMWriter to pcmWriter.
+	resampledPCMWriter := pcmWriter
+	if sourceSampleRate != DefaultOpusSampleRate {
+		resampledPCMWriter = media.ResampleWriter(pcmWriter, sourceSampleRate)
+	}
+
+	t := &PCMLocalTrack{
+		TrackLocalStaticSample: track,
+		opusWriter:             opusWriter,
+		pcmWriter:              pcmWriter,
+		resampledPCMWriter:     resampledPCMWriter,
+		sourceSampleRate:       sourceSampleRate,
+		frameDuration:          defaultPCMSampleDuration,
+		sourceChannels:         sourceChannels,
+		chunkBuffer:            new(deque.Deque[int16]),
+		chunksPerSample:        (sourceSampleRate * sourceChannels * int(defaultPCMSampleDuration/time.Nanosecond)) / 1e9,
+		writeSilenceOnNoData:   params.WriteSilenceOnNoData,
+	}
+
+	t.cond = sync.NewCond(&t.mu)
+
+	go t.processSamples()
+	return t, nil
+}
+
+func (t *PCMLocalTrack) pushChunksToBuffer(sample media.PCM16Sample) {
+	for _, chunk := range sample {
+		t.chunkBuffer.PushBack(chunk)
+	}
+}
+
+func (t *PCMLocalTrack) waitUntilBufferHasChunks(count int) bool {
+	var didWait bool
+
+	for t.chunkBuffer.Len() < count && !t.closed.Load() {
+		t.cond.Wait()
+		didWait = true
+	}
+
+	return didWait
+}
+
+func (t *PCMLocalTrack) getChunksFromBuffer() (media.PCM16Sample, bool) {
+	chunks := make(media.PCM16Sample, t.chunksPerSample)
+
+	var didWait = false
+	if !t.writeSilenceOnNoData {
+		didWait = t.waitUntilBufferHasChunks(t.chunksPerSample)
+	}
+
+	if t.closed.Load() {
+		return nil, false
+	}
+
+	for i := 0; i < t.chunksPerSample; i++ {
+		if t.chunkBuffer.Len() == 0 {
+			// this will zero-init at index i, which will be a silent chunk.
+			// if writeSilenceOnNoData is false, this condition will never be true.
+			continue
+		} else {
+			chunks[i] = t.chunkBuffer.PopFront()
+		}
+	}
+
+	return chunks, didWait
+}
+
+func (t *PCMLocalTrack) WriteSample(sample media.PCM16Sample) error {
+	if t.closed.Load() {
+		return errors.New("track is closed")
+	}
+
+	t.mu.Lock()
+	t.pushChunksToBuffer(sample)
+	t.cond.Broadcast()
+	t.mu.Unlock()
+	return nil
+}
+
+func (t *PCMLocalTrack) processSamples() {
+	ticker := time.NewTicker(t.frameDuration)
+	defer ticker.Stop()
+
+	for {
+		if t.closed.Load() {
+			break
+		}
+
+		t.mu.Lock()
+		sample, didWait := t.getChunksFromBuffer()
+		if sample != nil {
+			// sample is only nil when the track is closed, so we don't need to
+			// adjust ticker for this case.
+			t.resampledPCMWriter.WriteSample(sample)
+			if didWait {
+				ticker.Reset(t.frameDuration)
+			}
+		}
+		t.mu.Unlock()
+		<-ticker.C
+	}
+}
+
+func (t *PCMLocalTrack) Close() {
+	firstClose := t.closed.CompareAndSwap(false, true)
+	// avoid closing the writer multiple times
+	if firstClose {
+		t.mu.Lock()
+		defer t.mu.Unlock()
+		t.chunkBuffer.Clear()
+		t.cond.Broadcast()
+		t.resampledPCMWriter.Close()
+		t.pcmWriter.Close()
+		t.opusWriter.Close()
+	}
+}
+
+type PCMRemoteTrackParams struct {
```

Comment:
:nit: not needed now, my preference is to make a separate file for each struct. This can be split into pcmlocaltrack.go and pcmremotetrack.go